File: solutions/cpp/space-age/1/space_age.cpp
#include "space_age.h"

namespace space_age
{
    space_age::space_age(long seconds)
    {
        this->_seconds = seconds;
    }

    long space_age::seconds() const
    {
        return this->_seconds;
    }

    double space_age::on_mercury() const
    {
        return (_seconds / SECONDS_IN_YEAR) / MERCURY;
    }

    double space_age::on_venus() const
    {
        return (_seconds / SECONDS_IN_YEAR) / VENUS;
    }

    double space_age::on_earth() const
    {
        return (_seconds / SECONDS_IN_YEAR) / EARTH;
    }

    double space_age::on_mars() const
    {
        return (_seconds / SECONDS_IN_YEAR) / MARS;
    }

    double space_age::on_jupiter() const
    {
        return (_seconds / SECONDS_IN_YEAR) / JUPITER;
    }

    double space_age::on_saturn() const
    {
        return (_seconds / SECONDS_IN_YEAR) / SATURN;
    }

    double space_age::on_uranus() const
    {
        return (_seconds / SECONDS_IN_YEAR) / URANUS;
    }

    double space_age::on_neptune() const
    {
        return (_seconds / SECONDS_IN_YEAR) / NEPTUNE;
    }
} // namespace space_age

File: solutions/haskell/space-age/1/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

secondsInAYear = 31557600
data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn planet seconds = (seconds / secondsInAYear) / (getPlanetaryYear planet)

getPlanetaryYear :: Planet -> Float
getPlanetaryYear Mercury = 0.2408467
getPlanetaryYear Venus = 0.61519726
getPlanetaryYear Earth = 1.0 -- we're home
getPlanetaryYear Mars =  1.8808158
getPlanetaryYear Jupiter = 11.862615
getPlanetaryYear Saturn = 29.447498
getPlanetaryYear Uranus = 84.016846
getPlanetaryYear Neptune = 164.79132